system/include/emscripten/threading.h

```diff
@@ -299,6 +299,54 @@ int emscripten_dispatch_to_thread_async_(pthread_t target_thread,
   emscripten_dispatch_to_thread_async_(                                        \
     (target_thread), (sig), (void*)(func), (satellite), ##__VA_ARGS__)
 
+// Similar to emscripten_dispatch_to_thread, but waits on the dispatching thread
+// until `func` has returned on the target thread.
+int emscripten_dispatch_to_thread_sync_ptr(pthread_t target_thread,
+                                           void (*func)(void*),
+                                           void* arg);
+int emscripten_dispatch_to_thread_sync_args(pthread_t target_thread,
+                                            EM_FUNC_SIGNATURE sig,
+                                            void* func,
+                                            void* satellite,
+                                            va_list args);
+int emscripten_dispatch_to_thread_sync_(pthread_t target_thread,
+                                        EM_FUNC_SIGNATURE sig,
+                                        void* func,
+                                        void* satellite,
+                                        ...);
+#define emscripten_dispatch_to_thread_sync(                                    \
+  target_thread, sig, func, satellite, ...)                                    \
+  emscripten_dispatch_to_thread_sync_(                                         \
+    (target_thread), (sig), (void*)(func), (satellite), ##__VA_ARGS__)
+
+// Similar to emscripten_dispatch_to_thread, but waits on the dispatching thread
+// until `emscripten_async_as_sync_finish` is called on the
+// `em_async_as_sync_ctx` context pointer supplied to `func`. The first argument
```

Comment:
The existing `emscripten_dispatch_to_thread_sync()` function allows one to dispatch a task to another thread, and the calling/dispatching thread synchronously waits until the other thread runs the function and **returns** from that function call.

If I understand correctly, the idea with `async_as_sync` is to be able to delay the finish condition of the dispatched task? I.e. instead of the event of **returning** from the dispatched function signaling completion, the target thread will instead some time later manually call `emscripten_async_as_sync_finish()` to mark the dispatched call as finished? (to accommodate for async continuations from event handlers?)

If so, there does not seem to be a need to bake this feature as built-in to the proxying queue itself? It does have an effect of complicating the API quite a bit.

The major complication is the question about what should happen with the queue while there are one or more of these kinds of long-running async tasks pending? Should the queue retain them until they are asynchronously signaled to be finished? Or should they be removed from the async call queue already before they are signaled to be finished?

Users should be able to manually implement this kind of delayed completion signaling on top of synchronous or asynchronous dispatches via code like

```c
int longRunningAsyncTaskOnCalledThread(int *notifyAddressOnCompletion)
{
   // do long running async stuff, capture notifyAddressOnCompletion somewhere.
   // ...
   // done: wake all waiters.
   __atomic_store_n(notifyAddressOnCompletion, 1, __ATOMIC_SEQ_CST);
   emscripten_futex_wake(notifyAddressOnCompletion, INT_MAX);
}

int callingThread()
{
  int addressNotifiedOnCompletion = 0;
  emscripten_dispatch_to_thread(target_thread, EM_FUNC_SIG_VI, longRunningAsyncTaskOnCalledThread, 0, &addressNotifiedOnCompletion);
  emscripten_futex_wait(&addressNotifiedOnCompletion, 0, INFINITY);
}
```
i.e. users can get this behavior through the call queue, so this kind of long running async task could be treated orthogonally and would not need to be a feature of the dispatch queue itself?

Reply:
Yes, that's exactly right. OTOH, the normal `sync` version of the API can be built on top of the `async` version of the API in the same way, but we still consider it useful enough to provide as a utility. It makes sense to provide a utility to satisfy this async use case as well rather than forcing each user to implement it themselves. There is no appreciable complexity cost due to the layered nature of the implementation.